### DISCUSSION TITLE
Trying to find a nice way to declare actions

### DISCUSSION THREAD

**Discussion by jsruinedmylife** [statelyai/xstate#3812] [General]
When actions are declared as property of machine configuration, they will be executed when the machine will be restored from persisted state as expected

```js
const testMachine = createMachine(
  {
    id: 'test',
    initial: 'normal',
    states: {
      normal: {
        entry: [
          'activate'
        ]
      },
    }
  },
  {
    actions: {
      activate: (context, event) => {
        console.log('activating...');
      },
    }
  }
);
```
but in above case we can end up with a lot of lines of code in one single file.

As making current state being possible to store for example in local storage means transform the current state into JSON object, obviously a reference to variables will be gone, so in below case after restoring state machine action `activate` won't be executed.

```js
const activate = () => {
  console.log('activating...');
}

const testMachine = createMachine(
  {
    id: 'test',
    initial: 'normal',
    states: {
      normal: {
        entry: [
          activate // not as string
        ]
      },
    }
  }
);
```

It can be solved by using machine configuration actions properties as proxies to accurate functions
```js
// activate.js
export const logActivate = () => {
  console.log('activating...');
}

//machine.js
import { logActivate } from './activate.js'

const testMachine = createMachine(
  {
    id: 'test',
    initial: 'normal',
    states: {
      normal: {
        entry: [
          'activate'
        ]
      },
    }
  },
  {
    actions: {
      activate: (context, event) => {
        logActivate()
      },
    }
  }
);
```
but I'm not sure if this solution is well enough.

Do you have any suggestions on that?

**Comment by Andarist** [comment]
> When actions are declared as property of machine configuration, they will be executed when the machine will be restored from persisted state as expected

Out of curiosity - as we are reevaluating some behavior - why this is expected? An argument can be made that if you restore a configuration then those actions were already executed so why would you like to execute them **again**?

> It can be solved by using machine configuration actions properties as proxies to accurate functions

This works if you want to declare your functions outside of the machine. It allows your functions to be less coupled to XState (as you might map the context/event/etc to some other things and use those as arguments).
